src/ucp/rndv/rndv_mtype.inl

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2021, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCP_RNDV_MTYPE_INL_
+#define UCP_RNDV_MTYPE_INL_
+
+#include "proto_rndv.inl"
+
+#include <ucp/core/ucp_worker.h>
+
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_proto_rndv_mtype_init(const ucp_proto_init_params_t *init_params,
+                          ucp_md_map_t *mdesc_md_map_p, size_t *frag_size_p)
+{
+    ucp_worker_h worker = init_params->worker;
+    ucs_status_t status;
+
+    if ((init_params->select_param->dt_class != UCP_DATATYPE_CONTIG) ||
+        (worker->mem_type_ep[init_params->select_param->mem_type] == NULL)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if (mdesc_md_map_p != NULL) {
+        status = ucp_mm_get_alloc_md_map(worker->context, mdesc_md_map_p);
+        if (status != UCS_OK) {
+            return status;
+        }
+    }
+
+    *frag_size_p = worker->context->config.ext.rndv_frag_size;
+    return UCS_OK;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_proto_rndv_mtype_request_init(ucp_request_t *req)
+{
+    ucp_worker_h worker = req->send.ep->worker;
+
+    req->send.rndv.mdesc = ucp_worker_mpool_get(&worker->rndv_frag_mp);
+    if (req->send.rndv.mdesc == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    return UCS_OK;
+}
+
+static UCS_F_ALWAYS_INLINE uct_mem_h
+ucp_proto_rndv_mtype_get_memh(ucp_mem_desc_t *mdesc, ucp_rsc_index_t memh_index)
+{
+    if (memh_index == UCP_NULL_RESOURCE) {
+        return UCT_MEM_HANDLE_NULL;
+    }
+
+    ucs_assertv(memh_index < ucs_popcount(mdesc->memh->md_map),
+                "memh_index=%d md_map=0x%" PRIx64, memh_index,
+                mdesc->memh->md_map);
+    return mdesc->memh->uct[memh_index];
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_proto_rndv_mtype_next_iov(ucp_request_t *req,
+                              const ucp_proto_rndv_bulk_priv_t *rpriv,
+                              const ucp_proto_multi_lane_priv_t *lpriv,
+                              ucp_datatype_iter_t *next_iter, uct_iov_t *iov)
+{
+    ucp_mem_desc_t *mdesc = req->send.rndv.mdesc;
+    size_t max_payload    = ucp_proto_rndv_bulk_max_payload(req, rpriv, lpriv);
+
+    ucs_assert(req->send.state.dt_iter.dt_class == UCP_DATATYPE_CONTIG);
+
+    iov->length = ucp_datatype_iter_next(&req->send.state.dt_iter, max_payload,
+                                         next_iter);
+    iov->buffer = UCS_PTR_BYTE_OFFSET(mdesc + 1,
+                                      req->send.state.dt_iter.offset);
+    iov->memh   = ucp_proto_rndv_mtype_get_memh(mdesc, lpriv->super.memh_index);
+    iov->stride = 0;
+    iov->count  = 1;
```

Comment:
can be the common function with the code below?
```
void ucp_proto_rndv_mtype_iov_setup(uct_iov_t *iov, uct_mem_h memh, ucp_mem_desc_t *mdesc, size_t length, size_t offset)
{
    iov->length = length;
    iov->buffer = UCS_PTR_BYTE_OFFSET(mdesc + 1, offset);
    iov->memh   = memh;
    iov->stride = 0;
    iov->count  = 1;
}
```
and use it:
1. `ucp_proto_rndv_mtype_next_iov()`
```
ucp_proto_rndv_mtype_iov_setup(iov,
                               ucp_proto_rndv_mtype_get_memh(mdesc,
                                                             lpriv->super.memh_index),
                               mdesc,
                               ucp_datatype_iter_next(&req->send.state.dt_iter,
                                                      max_payload,
                                                      next_iter),
                               req->send.state.dt_iter.offset);
```
2. `ucp_proto_rndv_mtype_copy()`
```
ucp_proto_rndv_mtype_iov_setup(&iov, ucp_memh2uct(mdesc->memh, md_index), mdesc,
                               req->send.state.dt_iter.length, 0);
```